File: BraintreePaymentFlow/LocalPayment/BTLocalPaymentResult.m
#import "BTLocalPaymentResult.h"

#import <BraintreeCore/BTPostalAddress.h>

@implementation BTLocalPaymentResult

- (instancetype)initWithNonce:(NSString *)nonce
                  description:(NSString *)description
                         type:(NSString *)type
                        email:(NSString *)email
                    firstName:(NSString *)firstName
                     lastName:(NSString *)lastName
                        phone:(NSString *)phone
               billingAddress:(BTPostalAddress *)billingAddress
              shippingAddress:(BTPostalAddress *)shippingAddress
             clientMetadataId:(NSString *)clientMetadataId
                      payerId:(NSString *)payerId
{
    if (self = [super init]) {
        _nonce = nonce;
        _localizedDescription = description;
        _type = type;
        _email = email;
        _firstName = firstName;
        _lastName = lastName;
        _phone = phone;
        _billingAddress = [billingAddress copy];
        _shippingAddress = [shippingAddress copy];
        _clientMetadataId = clientMetadataId;
        _payerId = payerId;
    }
    return self;
}

@end

File: Sources/BraintreeLocalPayment/BTLocalPaymentResult.swift
import Foundation

#if canImport(BraintreeCore)
import BraintreeCore
#endif

@objcMembers public class BTLocalPaymentResult: NSObject {
    
    /// The billing address.
    public let billingAddress: BTPostalAddress?
    
    /// Client Metadata ID associated with this transaction.
    public let clientMetadataID: String?
    
    /// Payer's email address.
    public let email: String?
    
    /// Payer's first name.
    public let firstName: String?
    
    /// Payer's last name.
    public let lastName: String?
    
    /// The one-time use payment method nonce.
    public let nonce: String
    
    /// Payer ID associated with this transaction.
    public let payerID: String?
    
    /// Payer's phone number.
    public let phone: String?
    
    /// The shipping address.
    public let shippingAddress: BTPostalAddress?
    
    /// The type of the tokenized payment.
    public let type: String?
    
    /// :nodoc:
    @_documentation(visibility: private)
    init?(json: BTJSON) {
        let paypalAccount = json["paypalAccounts"][0]
        
        guard let nonce = paypalAccount["nonce"].asString() else { return nil }
        self.nonce = nonce
        
        type = paypalAccount["type"].asString()
        
        let details = paypalAccount["details"]
        clientMetadataID = details["correlationId"].asString()
        
        if details["payerInfo"]["email"].isString {
            email = details["payerInfo"]["email"].asString()
        } else {
            email = details["email"].asString()
        }
        
        firstName = details["payerInfo"]["firstName"].asString()
        lastName = details["payerInfo"]["lastName"].asString()
        phone = details["payerInfo"]["phone"].asString()
        payerID = details["payerInfo"]["payerId"].asString()
        
        if let payerShippingAddress = details["payerInfo"]["shippingAddress"].asAddress() {
            shippingAddress = payerShippingAddress
        } else {
            shippingAddress = details["payerInfo"]["accountAddress"].asAddress()
        }
        
        billingAddress = details["payerInfo"]["shippingAddress"].asAddress()
    }
}
